### PR TITLE
[Core] Continuous distance voting for undefined raycasting

### PR DESCRIPTION
**Pull request by rubenzorrilla**
Predator is back!

With this PR we add a voting technique similar to the one that we had in the past. Thus, when the raycasting is not consistent (all the rays doesn't yield the same colour), we cast some extra rays (4 in 2D and 8 in 3D), which origin is the same as the main one but perturbed a +/- epsilon value, to get more colour values. With this colour values we vote to finally decide if the origin node is in/out.

Note that this is rarely done (e.g. if no intersection is found because the ray intersects through a node and similar situations). In a 200k nodes mesh, only 20 nodes were not consistent. However, it improves quite a lot the quality of the obtained distance field.

### CODE REVIEW

File: kratos/processes/calculate_distance_to_skin_process.cpp [hunk 1]
```diff
@@ -236,8 +234,20 @@ namespace Kratos
...
+				n_ray_neg++;
+			} else {
+				n_ray_pos++;
+			}
         }
 
+		// Check the obtained cartesian ray colors
+		// If this situation happens, do the "evolved Predator" raycasting to vote
+		if (n_ray_neg != 0 && n_ray_pos != 0) {
+			const double ray_epsilon = 1e-4;
```
Comment (rubenzorrilla) [inline kratos/processes/calculate_distance_to_skin_process.cpp]: @roigcarlo  @pooyan-dadvand @RiccardoRossi Guys, what about this epsilon? Is the one used to perturb the origin ray. Maybe it's a good idea to create an extra constructor with it and if it is not provided it takes a default value.

Comment (pooyan-dadvand) [reply kratos/processes/calculate_distance_to_skin_process.cpp]: Is it in normalize space (key) or in real space (coordinate) dimension? 

In my opinion this epsilon should be very small. Much smaller than this.... But a more than the epsilon of intersection test.

Comment (RiccardoRossi) [reply kratos/processes/calculate_distance_to_skin_process.cpp]: in my opinion the epsilon should be either very small or relative to a given measure of the size...

one option is to expose it as a parameter to the user

Comment (pooyan-dadvand) [reply kratos/processes/calculate_distance_to_skin_process.cpp]: I would go for one default value and expose to change by user...

Comment (rubenzorrilla) [reply kratos/processes/calculate_distance_to_skin_process.cpp]: OK! I also like to expose it to the user. To have a relative calculation for such epsilon might lead to unexpected behaviours.

**Comment by rubenzorrilla** [comment]
I will take care of this ASAP. I'm still working on the embedded FSI trying to figure out why the distance is not well computed for an specific case. If it something related to this, I will implement the changes in here.

File: kratos/processes/calculate_distance_to_skin_process.cpp [hunk 2]
```diff
@@ -194,20 +194,18 @@ namespace Kratos
...
 	{
-		typedef Element::GeometryType intersection_geometry_type;
-        typedef std::vector<std::pair<double, intersection_geometry_type*> > intersections_container_type;
-
         const double epsilon = 1e-12;
 		array_1d<double,TDim> distances;
-        intersections_container_type intersections;
+		unsigned int n_ray_pos(0), n_ray_neg(0);
+        IntersectionsContainerType intersections;
+		const array_1d<double,3> coords = rNode.Coordinates();
```
Comment (RiccardoRossi) [inline kratos/processes/calculate_distance_to_skin_process.cpp]: do you need to make a copyhere? isn't a reference enough?

**Comment by RiccardoRossi** [comment]
@rubenzorrilla i am not sure if you did address the comments already. At first sight this looks correct to me...

**Comment by rubenzorrilla** [comment]
@RiccardoRossi not yet... I'll let you know when I'm finished.

File: kratos/processes/calculate_distance_to_skin_process.cpp (hunk 2 above)
Comment (rubenzorrilla) [reply kratos/processes/calculate_distance_to_skin_process.cpp]: Not needed. Changed to reference.

File: kratos/processes/calculate_distance_to_skin_process.cpp (hunk 1 above)
Comment (rubenzorrilla) [reply kratos/processes/calculate_distance_to_skin_process.cpp]: Guys, I've just implemented a new constructor that allows the user to pass the extra rays epsilon.

**Comment by rubenzorrilla** [comment]
@RiccardoRossi finished!

File: kratos/processes/calculate_distance_to_skin_process.cpp [hunk 3]
```diff
@@ -194,21 +204,19 @@ namespace Kratos
 	template<std::size_t TDim>
 	double CalculateDistanceToSkinProcess<TDim>::DistancePositionInSpace(const Node<3> &rNode)
 	{
-		typedef Element::GeometryType intersection_geometry_type;
-        typedef std::vector<std::pair<double, intersection_geometry_type*> > intersections_container_type;
-
         const double epsilon = 1e-12;
 		array_1d<double,TDim> distances;
-        intersections_container_type intersections;
+		unsigned int n_ray_pos(0), n_ray_neg(0);
```
Comment (RiccardoRossi) [inline kratos/processes/calculate_distance_to_skin_process.cpp:209]: on line 207 i think you should be using the epsilon you passed in the constructor

Comment (rubenzorrilla) [reply kratos/processes/calculate_distance_to_skin_process.cpp:209]: This is the `epsilon`to decide what is 0 and what is not. It should be different, indeed, smaller, than the one in the constructor.

**Review by RiccardoRossi — APPROVED** [review]

**Pull request merged**